### PR TITLE
add contracts and test

### PR DESCRIPTION
**Pull request by soloviii**

### CODE REVIEW

File: Airdrop.sol [hunk 1]
```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import "./Token.sol";
```
Comment (BhHaipls) [inline Airdrop.sol]: use standard IERC20 for interact with token

File: Airdrop.sol [hunk 2]
```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import "./Token.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract Airdrop is Ownable{
+
+    Token public token;
```
Comment (BhHaipls) [inline Airdrop.sol]: use standard IERC20 for interact with token

File: Airdrop.sol [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import "./Token.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+contract Airdrop is Ownable{
+
+    Token public token;
+
+    constructor(address payable _token) {
+        token = Token(_token);
```
Comment (BhHaipls) [inline Airdrop.sol]: use standard IERC20 for interact with token

Comment (BhHaipls) [reply Airdrop.sol]: remove payable is not require

File: Airdrop.sol [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    function depositEther() external payable onlyOwner {
+
+    }
+
+    function dropEther(
+        address payable[] calldata _recipients,
+        uint256 _amount
+    ) external payable onlyOwner {
```
Comment (BhHaipls) [inline Airdrop.sol]: are you sure this function has recive  eth?

File: Airdrop.sol [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+        for(uint256 i; i < _recipients.length; i++){
+           _recipients[i].transfer(_amount);
```
Comment (BhHaipls) [inline Airdrop.sol]: transfer return true or false, need check this moment in contract

File: Airdrop.sol [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+        }
+    }
+
+    function updateTokenAddress(address payable _token) public onlyOwner {
+        token = Token(_token);
```
Comment (BhHaipls) [inline Airdrop.sol]: use standard IERC20 for interact with token

Comment (BhHaipls) [reply Airdrop.sol]: remove payable is not require

File: Airdrop.sol [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    function withdrawEther() external payable onlyOwner{
+        payable(msg.sender).transfer(address(this).balance);
+    }
+
+    function getTokenAddress() public view returns(address){
```
Comment (BhHaipls) [inline Airdrop.sol]: external

File: Airdrop.sol [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+        return address(token);
+    }
+
+    function depositTokens(address payable _token) public payable onlyOwner {
+        require(msg.value > 0, "ERC20: amount is not valid");
```
Comment (BhHaipls) [inline Airdrop.sol]: this function is incorrect, It should not be payable, and not check on msg.value ( is how much you send eth, not token)

File: Airdrop.sol [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+        
+        token = Token(_token);
```
Comment (BhHaipls) [inline Airdrop.sol]: you need deposite only tokens which you set in up

File: Airdrop.sol [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+        token.transferFrom(msg.sender,address(this), msg.value);
```
Comment (BhHaipls) [inline Airdrop.sol]: code style pleas, and transferFrom returns value (true/false) please read about it , and check return value

File: Airdrop.sol [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    function dropTokens(
+        address payable[] calldata _recipients,
```
Comment (BhHaipls) [inline Airdrop.sol]: payable not require

File: Airdrop.sol [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+        uint256 _amount
+    ) external payable onlyOwner {
```
Comment (BhHaipls) [inline Airdrop.sol]: payable not require (payable modifier need when you transfer eth to recipient/method

File: Airdrop.sol [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+        for(uint256 i; i < _recipients.length; i++){
+            token.transfer(_recipients[i], _amount);
```
Comment (BhHaipls) [inline Airdrop.sol]: I mentioned above

Comment (BhHaipls) [reply Airdrop.sol]: also you need check whether there is enough balance at the address of the contract to send the appropriate number of coins

File: Airdrop.sol [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+        }
+    }
+
+    function withdrawTokens(address payable _token) external payable onlyOwner{
```
Comment (BhHaipls) [inline Airdrop.sol]: payable not require

File: Airdrop.sol [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+        token = Token(_token);
```
Comment (BhHaipls) [inline Airdrop.sol]: You already set the token in the contract, use it

File: Airdrop.sol [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+
+        token.transfer(msg.sender, token.balanceOf(address(this)));
```
Comment (BhHaipls) [inline Airdrop.sol]: I mentioned above

**Review by BhHaipls — CHANGES_REQUESTED** [review]

File: Airdrop.sol (hunk 4 above)
Comment (BhHaipls) [reply Airdrop.sol]: remove payable, because you send ethe in depositEther function;
this modifier is not nessesary

File: Airdrop.sol [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+        address payable[] calldata _recipients,
+        uint256 _amount
+    ) external payable onlyOwner {
+        for(uint256 i; i < _recipients.length; i++){
+           (bool sent,) =  _recipients[i].call{value: _amount}("");
+           require(sent, "Failed to send Ether");
+        }
+    }
+
+    function updateTokenAddress(address _token) public onlyOwner {
```
Comment (BhHaipls) [inline Airdrop.sol]: Think about whether you need an external modifier here?

File: Airdrop.sol [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+    function withdrawEther() external payable onlyOwner{
+        payable(msg.sender).transfer(address(this).balance);
+    }
+
+    function getTokenAddress() external view returns(address){
+        return address(token);
+    }
+
+    function depositTokens(uint256 _amount) public onlyOwner {
```
Comment (BhHaipls) [inline Airdrop.sol]: Think about whether you need an external modifier here?

File: Airdrop.sol [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+    function withdrawEther() external payable onlyOwner{
+        payable(msg.sender).transfer(address(this).balance);
+    }
+
+    function getTokenAddress() external view returns(address){
+        return address(token);
+    }
+
+    function depositTokens(uint256 _amount) public onlyOwner {        
+        (bool sent) = token.transferFrom(msg.sender,address(this), _amount);
```
Comment (BhHaipls) [inline Airdrop.sol]: code style

File: Airdrop.sol [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+        for(uint256 i; i < _recipients.length; i++){
+            (bool sent) = token.transfer(_recipients[i], _amount);
+            require(sent, "Failed to drop tokens");
+        }
+    }
+
+    function withdrawTokens() external onlyOwner{
+        (bool sent) = token.transfer(msg.sender, token.balanceOf(address(this)));
+        require(sent, "Failed to withdraw tokens");
+    }
```
Comment (BhHaipls) [inline Airdrop.sol]: Also see the order of the functions https://docs.soliditylang.org/en/v0.8.9/style-guide.html

File: Airdrop.sol [hunk 21]
```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.9;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/access/Ownable.sol";
+
+contract Airdrop is Ownable{
```
Comment (BhHaipls) [inline Airdrop.sol]: AirDrop*

File: Airdrop.test.js [hunk 22]
```diff
@@ -0,0 +1,136 @@
...
+    .use(require('chai-bn')(BN))
+    .should();
+
+const Token = artifacts.require("Token");
+const Airdrop = artifacts.require("Airdrop");
+
+contract("Airdrop", ([owner, user1, user2, user3]) => {
+
+    beforeEach(async () => {
+        tokenInstance = await Token.new({ from: owner });
```
Comment (BhHaipls) [inline Airdrop.test.js]: {from: owner} is by default

File: Airdrop.test.js [hunk 23]
```diff
@@ -0,0 +1,136 @@
...
+        airdropInstance = await Airdrop.new(tokenInstance.address, { from: owner });
```
Comment (BhHaipls) [inline Airdrop.test.js]: {from: owner} is by default

File: Airdrop.test.js [hunk 24]
```diff
@@ -0,0 +1,136 @@
...
+    });
+
```
Comment (BhHaipls) [inline Airdrop.test.js]: use the following order in the tests:
describe("Function name", async() =>{
 it("", ...
 it("",...
}

File: Airdrop.test.js [hunk 25]
```diff
@@ -0,0 +1,136 @@
...
+    it("Should deposit ether to the contract", async () => {
+
+        await airdropInstance.depositEther({ from: owner, value: ether('2') });
+
+        let etherBalanceAirdrop = await balance.current(airdropInstance.address);
+
+        etherBalanceAirdrop.should.be.bignumber.equal(ether('2'));
+    });
+
```
Comment (BhHaipls) [inline Airdrop.test.js]: add next test for depositEther : "Should be fail if caller not owner"

Comment (BhHaipls) [reply Airdrop.test.js]: The procedure for checking the change in balance is as follows:
let beforeBalance = await balance.cu....
await operation....
let newBalance = await balance...
expect(beforeBalance).to.be.bignumber.equal(beforeBalance.add( "the amount that was transferred"));

Comment (BhHaipls) [reply Airdrop.test.js]: in order to avoid the situation that by this time the balance was already 2 ether

File: Airdrop.test.js [hunk 26]
```diff
@@ -0,0 +1,136 @@
...
+    it("Should update token for airdrop", async () => {
+
+        let newToken = await Token.new({ from: owner });
+        await airdropInstance.updateTokenAddress(newToken.address);
+
+        let actualTokenAddress = await airdropInstance.getTokenAddress();
+
+        actualTokenAddress.should.be.equal(newToken.address);
+    });
+
```
Comment (BhHaipls) [inline Airdrop.test.js]: add next test for updatetokenAddress: "Should be fail if caller not owner"

File: Airdrop.test.js [hunk 27]
```diff
@@ -0,0 +1,136 @@
...
+            afterDropBalanceUser1.should.be.bignumber.equal(beforeDropBalanceUser1.add(ether('1')));
+            afterDropBalanceUser2.should.be.bignumber.equal(beforeDropBalanceUser2.add(ether('1')));
+        });
+
+        it("Should expect revert drop ether", async () => {
+            await expectRevert(
+                airdropInstance.dropEther([user1, user2], ether('1'), { from: owner }),
+                "Failed to send Ether"
+            );
+        });
```
Comment (BhHaipls) [inline Airdrop.test.js]: add next test for dropEther: "Should be fail if caller not owner"

File: Airdrop.test.js [hunk 28]
```diff
@@ -0,0 +1,136 @@
...
+        let newToken = await Token.new({ from: owner });
+        await airdropInstance.updateTokenAddress(newToken.address);
+
+        let actualTokenAddress = await airdropInstance.getTokenAddress();
+
+        actualTokenAddress.should.be.equal(newToken.address);
+    });
+
+    describe("drop ether", async () => {
+
```
Comment (BhHaipls) [inline Airdrop.test.js]: Add test: Should be fail if array is empty
Add test: Should be fail if amount Is zero

**Review by BhHaipls — CHANGES_REQUESTED** [review]
